Implement localization as part of a mission

aeshub opened the pull request:
The localization procedure is now implemented and will run as part of regular mission scheduling.

If the queue is empty a localization mission will be started for the current deck. This assumes that the operator has confirmed that the robot is on the deck of the mission that has been scheduled.

If there is an existing mission the system will check if a new mission is in the same deck as that mission and if so schedule it. If not it will be rejected.

If the last mission finishes a return to home mission will be scheduled which puts the robot back at the default localization pose. If a mission is scheduled in between the return to home mission another localization will not be required.

aeshub committed:
Implement localization as part of a mission

The localization procedure is now implemented and will run as part of
regular mission scheduling.

If the queue is empty a localization mission will be started for the
current deck. This assumes that the operator has confirmed that the
robot is on the deck of the mission that has been scheduled.

If there is an existing mission the system will check if a new mission
is in the same deck as that mission and if so schedule it. If not it
will be rejected.

If the last mission finishes a return to home mission will be scheduled
which puts the robot back at the default localization pose. If a
mission is scheduled in between the return to home mission another
localization will not be required.

aeshub committed:
Use inspection target directly from Echo request

aeshub committed:
Use robot pose directly from Echo request

aeshub committed:
Use robot pose directly from Echo request

aeshub committed:
Override equals method for position object

aeshub committed:
Refactor orientation equality override

aeshub committed:
Rename file to EnuPosition

aeshub committed:
Format unnecessary lines

aeshub committed:
Add inspection target to custom mission query

aeshub committed:
Fix formatting

aeshub committed:
Fix formatting

Afonso-2403 commented on backend/api/Services/AreaService.cs:
```diff
@@ -61,65 +61,42 @@ public async Task<IEnumerable<Area>> ReadAll()
-                a.Name.ToLower().Equals(areaName.ToLower())
-            ).Include(a => a.SafePositions).Include(a => a.Installation)
-                .Include(a => a.Plant).Include(a => a.Deck).FirstOrDefaultAsync();
+            return await _context.Areas.Where(a => a.Installation.Id.Equals(installation.Id) && a.Name.ToLower().Equals(areaName.ToLower()))
```
Installation is nullable, right?
As we do elsewhere the null check shouldn't be removed

Afonso-2403 commented on backend/api/Services/LocalizationService.cs:
```diff
@@ -0,0 +1,230 @@
+            var missionInstallation = await _installationService.ReadByName(missionDefinition.InstallationCode);
+            var robotInstallation = await _installationService.ReadByName(robot.CurrentInstallation);
+
+            if (missionInstallation is null || robotInstallation is null)
```
This message might be misleading if the robot has no current Installation

Afonso-2403 commented on backend/api/Services/LocalizationService.cs:
```diff
@@ -0,0 +1,230 @@
+                _logger.LogError("{Message}", ErrorMessage);
+                throw new AreaNotFoundException(ErrorMessage);
+            }
+            if (currentRobotArea?.Deck is null)
```
Out of curiosity, do we need this null conditional operator, given that if it is null we already throw an exception?

Afonso-2403 commented on backend/api/Services/LocalizationService.cs:
```diff
@@ -0,0 +1,230 @@
+        public async Task<bool> RobotIsLocalized(string robotId)
+        {
+            var robot = await _robotService.ReadById(robotId);
+            if (robot != null)
```
Any reason for having the return inside the if instead of the throwing of the exception?
Given that in most places we do the opposite this looks less readable to me

prasm313 committed:
Correct mismatch between area.ts and arearesponse

UsamaEquinorAFK committed:
Add InspectionService

UsamaEquinorAFK committed:
Modify InspectionFindings table

UsamaEquinorAFK committed:
Add InspectionFindingsQuery

UsamaEquinorAFK committed:
Add InspectionFindingsController

UsamaEquinorAFK committed:
Add scope for InspectionFindingsService

prasm313 committed:
Use robotResponse in RobotController.cs

UsamaEquinorAFK committed:
Add InspectionFindings Response

andchiind committed:
Reduce dependency between react queue components

UsamaEquinorAFK committed:
Remove response file for inspectionfindings

tsundvoll committed:
Add new analytics user to the broker

andchiind committed:
Separate dialogs into different components

andchiind committed:
Separate components to new files

andchiind committed:
Simplify syntax

andchiind committed:
Remove unnecessary react state

andchiind committed:
Make autocompletes controlled components

andchiind committed:
Log post errors

andchiind committed:
Remove warnings

andchiind committed:
Display areas in decks when opening page

andchiind committed:
Split up large deck component

Eddasol committed:
Correct datatype in DeckResponse

Eddasol committed:
Set result type of deck to DeckResponse

andchiind committed:
Use context for tracking mission queue

aeshub committed:
Add migrations for inspection target

UsamaEquinorAFK committed:
Fix SECURITY.md

prasm313 committed:
Add missionRunResponse model to backend

Eddasol committed:
Fix filtering of missions by installation

Eddasol committed:
Fix mission placeholder

UsamaEquinorAFK committed:
Refactor after making inspection not owned

UsamaEquinorAFK committed:
Remove InspectionFindings table temporarily

andchiind committed:
Rename dialogs to avoid confusion

andchiind committed:
Removed duplicate button file

aeshub committed:
Add small fixes

UsamaEquinorAFK committed:
Add the inspectionfindings table back

prasm313 committed:
Use MissionRunResponse in MissionRunController.cs

prasm313 committed:
Add area column to history page

andchiind committed:
Remove warnings

UsamaEquinorAFK committed:
Refactor InspectionFindings controller

UsamaEquinorAFK committed:
Update InspectionFindings query

UsamaEquinorAFK committed:
Remove area from inspection findings model

UsamaEquinorAFK committed:
Refactor inspection findings service

andchiind committed:
Use missionId instead of id when comparing missions

Eddasol committed:
Correct return datatype of area endpoint

Eddasol committed:
Update datatype to arearesponse in area tests

andchiind committed:
Remove circular dependency in useEffect

prasm313 committed:
Use missionRunResponse in MissionRunService.cs

prasm313 committed:
Update MissionTests.cs with missionRunResponse

marianards committed:
Improve appearance of no missions available dialog

Christdej committed:
Update to net8

marianards committed:
Make add mission button and queue independent

UsamaEquinorAFK committed:
Use best practice for throwing exception according to CA2200

UsamaEquinorAFK committed:
Fix test errors in log methods

aeshub committed:
Refactor database operations to utilities class

aeshub committed:
Reintroduce skipped test

andchiind committed:
Readd populate db in missions test

Christdej committed:
Add eslint to frontend pipeline

eivindsjovold committed:
Align function definitions in FrontPage

andchiind committed:
Refactor pressure and battery display

Eddasol committed:
Make none HTML-element functions camel case

Eddasol committed:
Fix rerun mission

andchiind committed:
Add function to get user roles in controllers

Eddasol committed:
Make HTML-element functions pascal case

Eddasol committed:
Use labmda functions

andchiind committed:
Add access role check for DeckService

Eddasol committed:
Refactor InspectionRow

andchiind committed:
Add granular access to more endpoints

Also add access roles as migrations

andchiind committed:
Use signalr groups for granular access

Eddasol committed:
Rename components in TaskTable

Eddasol committed:
Fix initial render of missions

Eddasol committed:
Fix duplicated missions in mission queue

UsamaEquinorAFK committed:
Assign correct value to IsarStepId

UsamaEquinorAFK committed:
Set db context for InspectionFindings

because InspectionFindings is not an owned entity

UsamaEquinorAFK committed:
Add InspectionFindingService and set scope

UsamaEquinorAFK committed:
Add Event handler for inspection findings report

UsamaEquinorAFK committed:
Fix dotnet formating warnings

andchiind committed:
Add signalR groups for each role

Also add mocked tokens to make tests pass

Also made Installation an object ref in Robot

prasm313 committed:
Add json constructor for robotResponse.cs

UsamaEquinorAFK committed:
Refactor InspectionFindingService to do more processing in eventhandler.

eivindsjovold committed:
Change base image in frontend to alpine

andchiind committed:
Protect add, update and delete functions

andchiind committed:
Update prettier

prasm313 committed:
Refactor robotTest with robotResponse

aeshub committed:
Move mutex to event handler

UsamaEquinorAFK committed:
Fix naming convention for InspectionFinding

andchiind committed:
Fix test by limiting initDb to one installation

UsamaEquinorAFK committed:
Add GetMissionRun and MissionTask by IsarStepId in InspectionFindingService

UsamaEquinorAFK committed:
Generate customized string message to be sent via Teams API and add test for it

UsamaEquinorAFK committed:
Fix bug in InspectionFindingService to include plants

UsamaEquinorAFK committed:
Generate MSTeams AdaptiveCard

UsamaEquinorAFK committed:
Add function for sending generated report to Teams channel

UsamaEquinorAFK committed:
Refactor and cleanup InspectionFinding Service

andchiind committed:
Use huldra as testing facility

andchiind committed:
Make installation required in access roles

Afonso-2403 committed:
Add TODO comments to unhnadled exceptions

Afonso-2403 committed:
Refactor to remove dotnet format warnings

UsamaEquinorAFK committed:
Add trufflehog github workflow

UsamaEquinorAFK committed:
Add logging message to the start of InspectionFindingEventHandler

Afonso-2403 committed:
Refactor exceptions

Afonso-2403 committed:
Test localization functionality and fix failing tests

UsamaEquinorAFK committed:
Update logs in the inspectionfinding eventhandler

UsamaEquinorAFK committed:
Change interval to 10 secs in appsettings

UsamaEquinorAFK committed:
Remove project path dependency for fetching secret

UsamaEquinorAFK committed:
Update interval to 1 day

oysand committed:
Use inspection position from echo mission

oysand committed:
Collect mission areas

UsamaEquinorAFK committed:
Change interval in appsetting to debug service in dev

Afonso-2403 committed:
Change from objects to strings in CreateRobotQuery

oysand committed:
Verify single deck and select area for mission

The most prevalent area in the list will be selected for the mission

Afonso-2403 committed:
Test creating robot with a current area

andchiind committed:
Remove LINQ errors for filtering

tsundvoll committed:
Remove area from robot query

since area should only be set during locaization

marianards committed:
Improve mission queue placeholder

UsamaEquinorAFK committed:
Add cron expression for interval in InspectionFindingEventHandler

UsamaEquinorAFK committed:
Remove redundant delay interval

UsamaEquinorAFK committed:
Fix inspectionfinding test to become deterministic

tsundvoll committed:
Add Local appsettings

and disable Application Insights Telemetry locally and when debugging

oysand committed:
Add working area for huldra missions

oysand committed:
Update inspection page with new mission definition

oysand committed:
Support that echo test create mission definition

andchiind committed:
Require admin priviligies to view inspections

UsamaEquinorAFK committed:
Add logs for AdaptiveCard

Afonso-2403 committed:
Cancel missions if localization fails

Afonso-2403 committed:
Use Enum for MissionTask Type

UsamaEquinorAFK committed:
Pass NA in InspectionReport if some entities are null

UsamaEquinorAFK committed:
Remove writeline from test

UsamaEquinorAFK committed:
Remove useInMemory database from appsetting.Local

UsamaEquinorAFK committed:
Include inspections and tasks within missionRun call

UsamaEquinorAFK committed:
Update run time for service

aeshub committed:
Set default safe positions for areas

aeshub committed:
Remove redundant function

aeshub committed:
Simplify mission type checks

aeshub committed:
Ensure semaphore is released in case of exception

aeshub committed:
Use GetAreas function in AreaService

aeshub committed:
Fix formatting issues

aeshub committed:
Remove redundant services
